#include "mainwindow.h"
#include "ui_mainwindow.h"
#include "qaesencryption.h"

#include <QCryptographicHash>
#include <QDebug>

MainWindow::MainWindow(QWidget *parent)
    : QMainWindow(parent)
    , ui(new Ui::MainWindow)
{
    ui->setupUi(this);
}

MainWindow::~MainWindow()
{
    delete ui;
}

// üîê –ö–Ω–æ–ø–∫–∞: –∑–∞—à–∏—Ñ—Ä–æ–≤–∞—Ç—å
void MainWindow::on_pushButtonEncrypt_clicked()
{
    QString input = ui->lineEditInput->toPlainText().trimmed();
    QByteArray plainData = input.toUtf8();

    // –ö–ª—é—á: "superpassword" (–¥–æ–ø–æ–ª–Ω—è–µ–º –¥–æ 32 –±–∞–π—Ç)
    QByteArray password = "superpassword";
    password.resize(32, '\0');  // –¥–æ–ø–æ–ª–Ω–µ–Ω–∏–µ –Ω—É–ª—è–º–∏ –¥–æ 32 –±–∞–π—Ç
    QByteArray key = password;

    // ECB –Ω–µ –∏—Å–ø–æ–ª—å–∑—É–µ—Ç IV
    QAESEncryption encryption(QAESEncryption::AES_256, QAESEncryption::ECB, QAESEncryption::PKCS7);
    QByteArray encrypted = encryption.encode(plainData, key);

    QString base64 = encrypted.toBase64();
    ui->lineEditOutput->setPlainText(base64);

    qDebug() << "Input (hex):" << plainData.toHex();
    qDebug() << "Key (hex):" << key.toHex();
    qDebug() << "Encrypted (base64):" << base64;
}

// üîì –ö–Ω–æ–ø–∫–∞: —Ä–∞—Å—à–∏—Ñ—Ä–æ–≤–∞—Ç—å
void MainWindow::on_pushButtonDecrypt_clicked()
{
    QString inputBase64 = ui->lineEditOutput->toPlainText().trimmed();
    QByteArray encryptedData = QByteArray::fromBase64(inputBase64.toUtf8());

    QByteArray password = "superpassword";
    password.resize(32, '\0');  // –¥–æ–ø–æ–ª–Ω–µ–Ω–∏–µ –¥–æ 32 –±–∞–π—Ç
    QByteArray key = password;

    QAESEncryption encryption(QAESEncryption::AES_256, QAESEncryption::ECB, QAESEncryption::PKCS7);
    QByteArray decrypted = encryption.decode(encryptedData, key);

    // –£–¥–∞–ª–µ–Ω–∏–µ –ø–∞–¥–¥–∏–Ω–≥–∞ PKCS7
    if (!decrypted.isEmpty()) {
        char pad = decrypted.at(decrypted.size() - 1);
        int padVal = static_cast<unsigned char>(pad);
        if (padVal > 0 && padVal <= 16) {
            decrypted.chop(padVal);
        }
    }

    QString result = QString::fromUtf8(decrypted);
    ui->lineEditInput->setPlainText(result);

    qDebug() << "Decrypted (hex):" << decrypted.toHex();
}

void MainWindow::on_pushButtonHash_clicked()
{
    // –ü–æ–ª—É—á–∞–µ–º –≤–≤–æ–¥
    QString input = ui->lineEditHashInput->toPlainText().trimmed();
    QByteArray inputData = input.toUtf8();

    // –í—ã—á–∏—Å–ª—è–µ–º SHA-256
    QByteArray hash = QCryptographicHash::hash(inputData, QCryptographicHash::Sha256);

    // –ö–æ–Ω–≤–µ—Ä—Ç–∏—Ä—É–µ–º –≤ —Å—Ç—Ä–æ–∫—É hex
    QString hashHex = hash.toHex();

    // –ü–æ–∫–∞–∑—ã–≤–∞–µ–º —Ä–µ–∑—É–ª—å—Ç–∞—Ç –≤ –ø–æ–ª–µ –≤—ã–≤–æ–¥–∞
    ui->lineEditHashOutput->setPlainText(hashHex);

    // –û—Ç–ª–∞–¥–∫–∞
    qDebug() << "Input:" << input;
    qDebug() << "SHA-256 (hex):" << hashHex;
}
